AsyncDecodingSource.cpp

```diff
@@ -0,0 +1,408 @@
+}
+
+bool AsyncDecodingSource::configure(const sp<AMessage> format, const sp<Surface> surface, uint32_t flags) {
+    mCodec->setCallback(mNotify);
+    status_t err = mCodec->configure(format, surface, NULL /* crypto */, flags);
+    if (err != OK) {
+        ALOGE("Failed to configure codec '%s'", mComponentName.c_str());
+        return false;
+    }
+    mUsingSurface = surface != NULL;
+
+    // Set output format metadata
+    sp<AMessage> outputFormat;
+    mCodec->getOutputFormat(&outputFormat);
+    sp<MetaData> meta = new MetaData;
+    convertMessageToMetaData(outputFormat, meta);
+    mMeta.lock().set(meta);
+
+    mOutput.lock()->mReachedEOS = false;
+    ALOGI("Configured codec '%s'!", mComponentName.c_str());
+    return true;
+}
+
+status_t AsyncDecodingSource::start(MetaData *params) {
+    (void)params;
+    Mutexed<Output>::Locked me(mOutput);
+
+    if (mState != INIT) {
+        return -EINVAL;
+    }
+
+    status_t res = mCodec->start();
+    if (res == OK) {
+        res = mSource->start();
+    }
+
+    if (res == OK) {
+        mState = STARTED;
+        me->mReachedEOS = false;
+    } else {
+        mState = ERROR;
+    }
+    return res;
+}
+
+status_t AsyncDecodingSource::stop() {
+    Mutexed<Output>::Locked me(mOutput);
+    if (mState != STARTED) {
+        return -EINVAL;
+    }
+
+    // wait for any pending reads to complete
+    mState = STOPPING;
+    while (me->mReading) {
+        me.waitForCondition(me->mReadCondition);
+    }
+
+    status_t res1 = mCodec->stop();
+    if (res1 != OK) {
+        mCodec->release();
+    }
+    status_t res2 = mSource->stop();
+    if (res1 == OK && res2 == OK) {
+        mState = STOPPED;
+    } else {
+        mState = ERROR;
+    }
+    return res1 != OK ? res1 : res2;
+}
+
+sp<MetaData> AsyncDecodingSource::getFormat() {
+    Mutexed<sp<MetaData>>::Locked meta(mMeta);
+    return *meta;
+}
+
+AsyncDecodingSource::Output::Output()
+    : mReachedEOS(false),
+      mReading(false) {
+}
+
+status_t AsyncDecodingSource::read(
+#if ANDROID_MAJOR >= 9
+                    MediaBufferBase **buffer,
+#else
+                    MediaBuffer **buffer,
+#endif
+                    const ReadOptions *options) {
+    *buffer = NULL;
+    Mutexed<Output>::Locked me(mOutput);
+    me->mReading = true;
+    status_t res = OK;
+
+    if (mState != STARTED) {
+        return ERROR_END_OF_STREAM;
+    }
+
+    // flush codec on seek
+    MediaSource::ReadOptions::SeekMode mode;
+    int64_t timeUs = 0ll;
+    if (options != NULL && options->getSeekTo(&timeUs, &mode)) {
+        me->mReachedEOS = false;
+        mCodec->flush();
+    }
+
+    *buffer = NULL;
+    while (me->mBufferQueue.size() == 0 && !me->mReachedEOS) {
+        ALOGV("[%s] Waiting for output.", mComponentName.c_str());
+        me.waitForCondition(me->mAvailable);
+    }
+
+    if (mState == ERROR) {
+        res = ERROR;
+    }
+    else if (me->mBufferQueue.size() > 0) {
+        *buffer = *me->mBufferQueue.begin();
```

Comment:
Ignore at your leisure, and it's probably safest that you do. But I've seen a mix of `me->mBufferQueue.size() >/= 0` and `me->mBufferQueue.empty()`, personally (as in Qt isEmpty() is the prevailing idiom) empty() is a little easier to parse.

But assuming `List` is actually a type alias of std::vector or close enough you could call `me->mBufferQueue.front()` and eliminate the iterator dereference.

Reply:
There doesn't seem to be a front() in android::List, but I switched the size > 0 to !empty